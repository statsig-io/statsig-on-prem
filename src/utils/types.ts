export type OmitUndefined<T> = {
  [K in keyof T as undefined extends T[K] ? never : K]: T[K];
};
