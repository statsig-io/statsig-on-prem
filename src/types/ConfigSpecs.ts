export declare type ConfigSpecs = {
  dynamic_configs: Array<APIConfigSpec>;
  feature_gates: Array<APIConfigSpec>;
  layer_configs: Array<APIConfigSpec>;
  layers: Record<string, string[]>;
  hashed_sdk_keys_to_entities?: Record<string, APIEntityNames>;
  has_updates: boolean;
  time: number;
};

export type APIEntityNames = {
  gates: string[];
  configs: string[];
};

export enum APIConfigType {
  DYNAMIC_CONFIG = "dynamic_config",
  FEATURE_GATE = "feature_gate",
}

export enum APIEntityType {
  AUTOTUNE = "autotune",
  DYNAMIC_CONFIG = "dynamic_config",
  EXPERIMENT = "experiment",
  FEATURE_GATE = "feature_gate",
  HOLDOUT = "holdout",
  LAYER = "layer",
  SEGMENT = "segment",
}

export type APIConfigSpec = {
  name: string;
  type: APIConfigType;
  salt: string;
  defaultValue: boolean | Record<string, unknown>;
  enabled: boolean;
  rules: Array<APIConfigRule>;
  isDeviceBased: boolean;
  idType: string;
  entity: APIEntityType;
  isActive?: boolean;
};

export type APIConfigRule = {
  name: string;
  groupName?: string | null;
  passPercentage: number;
  conditions: Array<APIConfigCondition>;
  returnValue: boolean | Record<string, unknown>;
  id: string;
  salt: string;
  idType: string;
  isExperimentGroup?: boolean;
};

export type APIConfigCondition = {
  type: APIConditionType;
  targetValue: string[] | number[] | string | number | null;
  operator: APIOperatorType | null;
  field: APIFieldName | string | null;
  additionalValues: Record<string, string> | null;
  idType: string;
};

export enum APIConditionType {
  FALLBACK = "fallback",
  PUBLIC = "public",
  IP_BASED = "ip_based",
  UA_BASED = "ua_based",
  USER_FIELD = "user_field",
  PASS_GATE = "pass_gate",
  FAIL_GATE = "fail_gate",
  CURRENT_TIME = "current_time",
  ENVIRONMENT_FIELD = "environment_field",
  USER_BUCKET = "user_bucket",
  UNIT_ID = "unit_id",
  JAVASCRIPT = "javascript",
  MULTI_PASS_GATE = "multi_pass_gate",
  MULTI_FAIL_GATE = "multi_fail_gate",
  USER_BUCKET_ENCODED = "user_bucket_encoded",
}

export enum APIFieldName {
  BROWSER_NAME = "browser_name",
  BROWSER_VERSION = "browser_version",
  OS_NAME = "os_name",
  OS_VERSION = "os_version",

  // user data fields
  APP_VERISON = "appVersion",
  COUNTRY = "country",
  EMAIL = "email",
  USER_ID = "userID",
  IP = "ip",
  LOCALE = "locale",

  // environment fields
  TIER = "tier",
}

export enum APIOperatorType {
  ANY = "any",
  NONE = "none",

  GT = "gt",
  LT = "lt",
  LTE = "lte",
  GTE = "gte",

  VERSION_GT = "version_gt",
  VERSION_LT = "version_lt",
  VERSION_EQ = "version_eq",

  VERSION_GTE = "version_gte",
  VERSION_LTE = "version_lte",

  STR_CONTAINS_ANY = "str_contains_any",
  STR_CONTAINS_NONE = "str_contains_none",

  STR_STARTS_WITH_ANY = "str_starts_with_any",
  STR_ENDS_WITH_ANY = "str_ends_with_any",

  AFTER = "after",
  BEFORE = "before",
  ON = "on",

  IN_SEGMENT_LIST = "in_segment_list",
  NOT_IN_SEGMENT_LIST = "not_in_segment_list",

  EQ = "eq",
  NEQ = "neq",

  STR_MATCHES = "str_matches",

  ENCODED_ANY = "encoded_any",
}
