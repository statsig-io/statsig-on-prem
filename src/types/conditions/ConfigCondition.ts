import {
  APIConditionType,
  APIConfigCondition,
  APIFieldName,
  APIOperatorType,
} from "../ConfigSpecs";
import { z } from "zod";

const MultiTypes = z.union([
  z.string().array().nullable(),
  z.number().array().nullable(),
  z.null(),
]);

const SingleTypes = z.union([z.string(), z.number(), z.null()]);

const MultiValue = <T extends z.ZodTypeAny>(type: T) =>
  type.refine(
    (t) => MultiTypes.safeParse(t).success,
    `Expected array of strings or numbers.`
  );

const SingleValue = <T extends z.ZodTypeAny>(type: T) =>
  type.refine(
    (t) => SingleTypes.safeParse(t).success,
    "Expected string or number."
  );

export function SupportedOperations<
  O extends z.ZodTypeAny,
  V extends z.ZodTypeAny,
  F extends z.ZodTypeAny
>(operators: O, targetValue: V, field: F) {
  return z
    .union([
      z.object({
        operator: z.literal(APIOperatorType.ANY),
        targetValue: MultiValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.NONE),
        targetValue: MultiValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.EQ),
        targetValue: SingleValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.NEQ),
        targetValue: SingleValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.STR_CONTAINS_ANY),
        targetValue: z.string().array(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.STR_CONTAINS_NONE),
        targetValue: z.string().array(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.STR_MATCHES),
        targetValue: z.string(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.LT),
        targetValue: z.number(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.LTE),
        targetValue: z.number(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.GT),
        targetValue: z.number(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.GTE),
        targetValue: z.number(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.VERSION_LT),
        targetValue: SingleValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.VERSION_LTE),
        targetValue: SingleValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.VERSION_GT),
        targetValue: SingleValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.VERSION_GTE),
        targetValue: SingleValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.VERSION_EQ),
        targetValue: SingleValue(targetValue),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.ON),
        targetValue: z.number(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.AFTER),
        targetValue: z.number(),
        field: field,
      }),
      z.object({
        operator: z.literal(APIOperatorType.BEFORE),
        targetValue: z.number(),
        field: field,
      }),
    ])
    .refine(
      (
        data
      ): data is {
        operator: z.infer<typeof operators>;
        targetValue: z.infer<typeof targetValue>;
        field: z.infer<typeof field>;
      } => operators.safeParse(data.operator).success
    );
}

export enum ConfigConditionType {
  CUSTOM_FIELD = "custom_field",
  EMAIL = "email",
  LOCALE = "locale",
  UNIT_ID = "unit_id",
  USER_ID = "user_id",
}

export type ConfigConditionJSON = {
  type: ConfigConditionType;
  operator: APIOperatorType;
  field: string | APIFieldName | null;
  targetValue: string[] | number[] | string | number | null;
};

export default abstract class ConfigCondition {
  public abstract getConfigConditionType(): ConfigConditionType;
  public abstract getAPIConditionType(): APIConditionType;
  public abstract getIDType(): string;
  public abstract getOperator(): APIOperatorType;
  public abstract getField(): APIFieldName | string | null;
  public abstract getTargetValue():
    | string[]
    | number[]
    | string
    | number
    | null;

  // Schematized for SDK evaluation
  public getAPICondition(): APIConfigCondition {
    return {
      type: this.getAPIConditionType(),
      targetValue: this.getTargetValue(),
      operator: this.getOperator(),
      field: this.getField(),
      additionalValues: null,
      idType: this.getIDType(),
    };
  }

  // Schematized for storage
  public getConditionJSON(): ConfigConditionJSON {
    return {
      type: this.getConfigConditionType(),
      targetValue: this.getTargetValue(),
      operator: this.getOperator(),
      field: this.getField(),
    };
  }
}
