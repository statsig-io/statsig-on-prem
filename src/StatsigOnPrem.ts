import { StatsigInterface } from "./interfaces/StatsigInterface";
import { StorageInterface } from "./interfaces/StorageInterface";
import type {
  DynamicConfig,
  DynamicConfigCreationArgs,
  DynamicConfigUpdateArgs,
} from "./types/DynamicConfig";
import type {
  Experiment,
  ExperimentCreationArgs,
  ExperimentUpdateArgs,
} from "./types/Experiment";
import type {
  FeatureGate,
  FeatureGateCreationArgs,
  FeatureGateUpdateArgs,
} from "./types/FeatureGate";
import { EntityNames } from "./types/EntityNames";
import { GLOBAL_ASSOC_KEY } from "./utils/StorageUtils";
import { ConfigSpecs } from "./types/ConfigSpecs";
import ConfigSpecsUtils, { ConfigSpecsOptions } from "./utils/ConfigSpecsUtils";
import { filterNulls } from "./utils/filterNulls";
import EntityExperiment from "./entities/EntityExperiment";
import EntityDynamicConfig from "./entities/EntityDynamicConfig";
import EntityFeatureGate from "./entities/EntityFeatureGate";
import { IDUtils } from "./utils/IDUtils";
import { SpecsCacheInterface } from "./interfaces/SpecsCacheInterface";
import StorageHandler from "./utils/StorageHandler";
import { TargetAppNames } from "./types/TargetAppNames";
import CacheHandler from "./utils/CacheHandler";
import { SDKKeysCacheInterface } from "./interfaces/SDKKeyCacheInterface";
import ConfigRuleBuilder from "./utils/ConfigRuleBuilder";
import { genFilter } from "./utils/genFilter";

type Plugins = Partial<{
  specsCache: SpecsCacheInterface;
  sdkKeysCache: SDKKeysCacheInterface;
}>;

export default class StatsigOnPrem implements StatsigInterface {
  private store: StorageHandler;
  private cache: CacheHandler;
  public constructor(storage: StorageInterface, plugins?: Plugins) {
    this.store = new StorageHandler(storage);
    this.cache = new CacheHandler({
      specs: plugins?.specsCache,
      keys: plugins?.sdkKeysCache,
    });
  }

  public async initialize(): Promise<void> {
    await this.store.initialize();
  }

  public async shutdown(): Promise<void> {
    await this.store.shutdown();
  }

  public async clearCache(sdkKey?: string): Promise<void> {
    if (sdkKey) {
      await this.cache.clearSpecs(sdkKey);
    } else {
      await this.cache.clearAll();
    }
  }

  public async clearCacheForTargetApp(...targetApps: string[]): Promise<void> {
    await Promise.all(
      targetApps.map(async (targetApp) => {
        const sdkKeys = await this.store.getSDKKeysForTargetApp(targetApp);
        if (sdkKeys) {
          await this.cache.clearSpecs(...Array.from(sdkKeys));
        }
      })
    );
    const globalSDKKeys = await this.getGlobalSDKKeys();
    await this.cache.clearSpecs(...Array.from(globalSDKKeys));
  }

  public async clearCacheForEntity(
    entity: FeatureGate | Experiment | DynamicConfig
  ): Promise<void> {
    const targetApps = entity.targetApps;
    if (targetApps == null) {
      await this.cache.clearSpecs();
    } else {
      await this.clearCacheForTargetApp(...Array.from(targetApps));
    }
  }

  public async getConfigSpecs(
    sdkKey: string,
    options?: ConfigSpecsOptions
  ): Promise<ConfigSpecs> {
    const registeredKeys = await this.store.getRegisteredSDKKeys();
    if (!registeredKeys.has(sdkKey)) {
      console.warn("Attempting to use a non-registered key");
    }

    const cachedSpecs = await this.cache.getSpecs(sdkKey, options);
    if (cachedSpecs) {
      return cachedSpecs;
    }

    const targetApps = await this.store.getTargetAppsFromSDKKey(sdkKey);
    const entities = await this.store.getEntityAssocsForMultipleTargetApps(
      targetApps
    );
    if (entities == null) {
      return ConfigSpecsUtils.getEmptyConfigSpecs();
    }

    const gates = filterNulls(
      await Promise.all(
        Array.from(entities.gates).map((name) =>
          this.getGate(name).then((gate) =>
            gate ? new EntityFeatureGate(gate) : null
          )
        )
      )
    );
    const configs = filterNulls(
      await Promise.all(
        Array.from(entities.configs).map((name) =>
          this.getConfig(name).then((config) =>
            config ? new EntityDynamicConfig(config) : null
          )
        )
      )
    );
    const experiments = filterNulls(
      await Promise.all(
        Array.from(entities.experiments).map((name) =>
          this.getExperiment(name).then((exp) =>
            exp ? new EntityExperiment(exp) : null
          )
        )
      )
    );

    const configSpecs: ConfigSpecs = {
      feature_gates: gates.map((gate) => ConfigSpecsUtils.getConfigSpec(gate)),
      dynamic_configs: [...configs, ...experiments].map((config) =>
        ConfigSpecsUtils.getConfigSpec(config)
      ),
      layer_configs: [],
      layers: {},
      has_updates: true,
      time: Date.now(),
    };

    if (options?.ssr) {
      configSpecs.hashed_sdk_keys_to_entities =
        await ConfigSpecsUtils.getHashedSDKKeysToEntities(
          this.store,
          options?.ssr
        );
    }

    await this.cache.cacheSpecs(sdkKey, options, configSpecs);
    return configSpecs;
  }

  /* Feature Gate */

  public async getGate(name: string): Promise<FeatureGate | null> {
    return await this.store.getGate(name);
  }

  public async createGate(
    name: string,
    args: FeatureGateCreationArgs
  ): Promise<void> {
    const { targetApps, idType, ...metadata } = args;
    const gate: FeatureGate = {
      name,
      salt: IDUtils.generateNewSalt(),
      idType: idType ?? "userID",
      targetApps: new Set(targetApps ?? []),
      ...metadata,
    };
    await this.store.addGate(gate);
    await this.clearCacheForEntity(gate);
  }

  public async updateGate(
    name: string,
    args: FeatureGateUpdateArgs
  ): Promise<void> {
    const gate = await this.getGate(name);
    if (gate == null) {
      console.warn("Attempting to update non-existent gate");
      return;
    }
    await this.clearCacheForEntity(gate);
    await this.store.updateGate(gate, args);
    await this.clearCacheForEntity(gate);
  }

  public async deleteGate(name: string): Promise<void> {
    const gate = await this.store.getGate(name);
    if (!gate) {
      console.warn("Attempting to delete non-existent gate");
      return;
    }
    await this.store.removeGate(name);
    await this.clearCacheForEntity(gate);
  }

  public async addTargetAppsToGate(
    name: string,
    targetApps: string[]
  ): Promise<void> {
    const gate = await this.getGate(name);
    if (gate == null) {
      console.warn("Attempting to update non-existent gate");
      return;
    }
    await this.store.addTargetAppsToGate(gate, targetApps);
    await this.clearCacheForTargetApp(...targetApps);
  }

  public async removeTargetAppsFromGate(
    name: string,
    targetApps: string[]
  ): Promise<void> {
    const gate = await this.getGate(name);
    if (gate == null) {
      console.warn("Attempting to update non-existent gate");
      return;
    }
    await this.store.removeTargetAppsFromGate(gate, targetApps);
    await this.clearCacheForTargetApp(...targetApps);
  }

  /* Experiment */

  public async getExperiment(name: string): Promise<Experiment | null> {
    return await this.store.getExperiment(name);
  }

  public async createExperiment(
    name: string,
    args: ExperimentCreationArgs
  ): Promise<void> {
    const { targetApps, idType, ...metadata } = args;
    const experiment: Experiment = {
      name,
      salt: IDUtils.generateNewSalt(),
      idType: idType ?? "userID",
      targetApps: new Set(targetApps ?? []),
      ...metadata,
    };
    await this.store.addExperiment(experiment);
    await this.clearCacheForEntity(experiment);
  }

  public async updateExperiment(
    name: string,
    args: ExperimentUpdateArgs
  ): Promise<void> {
    const experiment = await this.getExperiment(name);
    if (experiment == null) {
      console.warn("Attempting to update non-existent experiment");
      return;
    }
    if (experiment.started) {
      console.warn(
        "Attempting to update an experiment that has already started"
      );
      return;
    }
    await this.clearCacheForEntity(experiment);
    await this.store.updateExperiment(experiment, args);
    await this.clearCacheForEntity(experiment);
  }

  public async deleteExperiment(name: string): Promise<void> {
    const experiment = await this.store.getExperiment(name);
    if (!experiment) {
      console.warn("Attempting to delete non-existent experiment");
      return;
    }
    await this.store.removeExperiment(name);
    await this.clearCacheForEntity(experiment);
  }

  public async startExperiment(name: string): Promise<void> {
    const experiment = await this.getExperiment(name);
    if (experiment == null) {
      console.warn("Attempting to start an experiment that doesn't exist");
      return;
    }
    if (experiment.started) {
      console.warn(
        "Attempting to start an experiment that has already started"
      );
      return;
    }
    await this.store.updateExperiment(experiment, { started: true });
    await this.clearCacheForEntity(experiment);
  }

  public async addTargetAppsToExperiment(
    name: string,
    targetApps: string[]
  ): Promise<void> {
    const experiment = await this.getExperiment(name);
    if (experiment == null) {
      console.warn("Attempting to update non-existent experiment");
      return;
    }
    await this.store.addTargetAppsToExperiment(experiment, targetApps);
    await this.clearCacheForTargetApp(...targetApps);
  }

  public async removeTargetAppsFromExperiment(
    name: string,
    targetApps: string[]
  ): Promise<void> {
    const experiment = await this.getExperiment(name);
    if (experiment == null) {
      console.warn("Attempting to update non-existent experiment");
      return;
    }
    await this.store.removeTargetAppsFromExperiment(experiment, targetApps);
    await this.clearCacheForTargetApp(...targetApps);
  }

  /* Dynamic Config */

  public async getConfig(name: string): Promise<DynamicConfig | null> {
    return await this.store.getConfig(name);
  }

  public async createConfig(
    name: string,
    args: DynamicConfigCreationArgs
  ): Promise<void> {
    const { targetApps, idType, ...metadata } = args;
    const config: DynamicConfig = {
      name,
      salt: IDUtils.generateNewSalt(),
      idType: idType ?? "userID",
      targetApps: new Set(targetApps ?? []),
      ...metadata,
    };
    await this.store.addConfig(config);
    await this.clearCacheForEntity(config);
  }

  public async updateConfig(
    name: string,
    args: DynamicConfigUpdateArgs
  ): Promise<void> {
    const config = await this.getConfig(name);
    if (config == null) {
      console.warn("Attempting to update non-existent config");
      return;
    }
    const { targetApps, patchRules, ...changes } = args;
    const updated: DynamicConfig = { ...config, ...changes };
    if (targetApps) {
      await this.clearCacheForTargetApp(...targetApps);
    }
    if (patchRules) {
      const builder = new ConfigRuleBuilder(config);
      updated.rulesJSON = patchRules(builder).getRulesJSON();
    }
    await this.store.updateConfig(updated, args);
    await this.clearCacheForEntity(config);
  }

  public async deleteConfig(name: string): Promise<void> {
    const config = await this.getConfig(name);
    if (!config) {
      console.warn("Attempting to delete non-existent config");
      return;
    }
    await this.store.removeConfig(name);
    await this.clearCacheForEntity(config);
  }

  public async addTargetAppsToConfig(
    name: string,
    targetApps: string[]
  ): Promise<void> {
    const config = await this.getConfig(name);
    if (config == null) {
      console.warn("Attempting to update non-existent config");
      return;
    }
    await this.store.addTargetAppsToConfig(config, targetApps);
    await this.clearCacheForTargetApp(...targetApps);
  }

  public async removeTargetAppsFromConfig(
    name: string,
    targetApps: string[]
  ): Promise<void> {
    const config = await this.getConfig(name);
    if (config == null) {
      console.warn("Attempting to update non-existent config");
      return;
    }
    await this.store.removeTargetAppsFromConfig(config, targetApps);
    await this.clearCacheForTargetApp(...targetApps);
  }

  /* Target App */

  public async createTargetApp(
    name: string,
    entities: EntityNames
  ): Promise<void> {
    if (name === GLOBAL_ASSOC_KEY) {
      console.error("Invalid name for target app");
      return;
    }
    const existing = await this.store.getEntityAssocs(name);
    if (existing != null) {
      console.warn(`Target app already exists with the name ${name}`);
      return;
    }
    await this.store.addTargetApp(name, entities);
    await this.clearCacheForTargetApp(name);
  }

  public async updateTargetApp(
    name: string,
    entities: Partial<EntityNames>
  ): Promise<void> {
    const existing = await this.store.getEntityAssocs(name);
    if (existing == null) {
      console.warn("Attempting to update non-existent target app");
      return;
    }
    await this.store.updateTargetApp(name, { ...existing, ...entities });
    await this.clearCacheForTargetApp(name);
  }

  public async deleteTargetApp(name: string): Promise<void> {
    await this.store.removeTargetApp(name);
    await this.clearCacheForTargetApp(name);
  }

  public async addEntitiesToTargetApp(
    targetApp: string,
    entities: EntityNames
  ): Promise<void> {
    await this.store.addEntityAssocs(entities, targetApp);
    await Promise.all([
      ...Array.from(entities.gates).map((gate) =>
        this.addTargetAppsToGate(gate, [targetApp])
      ),
      ...Array.from(entities.configs).map((config) =>
        this.addTargetAppsToConfig(config, [targetApp])
      ),
      ...Array.from(entities.experiments).map((experiment) =>
        this.addTargetAppsToExperiment(experiment, [targetApp])
      ),
    ]);
    await this.clearCacheForTargetApp(targetApp);
  }

  public async removeEntitiesFromTargetApp(
    targetApp: string,
    entities: EntityNames
  ): Promise<void> {
    await this.store.removeEntityAssocs(entities, targetApp);
    await Promise.all([
      ...Array.from(entities.gates).map((gate) =>
        this.removeTargetAppsFromGate(gate, [targetApp])
      ),
      ...Array.from(entities.configs).map((config) =>
        this.removeTargetAppsFromConfig(config, [targetApp])
      ),
      ...Array.from(entities.experiments).map((experiment) =>
        this.removeTargetAppsFromExperiment(experiment, [targetApp])
      ),
    ]);
    await this.clearCacheForTargetApp(targetApp);
  }

  public async assignTargetAppsToSDKKey(
    targetApps: string[],
    sdkKey: string
  ): Promise<void> {
    await this.store.assignTargetAppsToSDKKey(targetApps, sdkKey);
    await this.cache.clearSpecs(sdkKey);
    await this.cache.clearGlobalSDKKeys();
  }

  public async removeTargetAppsFromSDKKey(
    targetApps: string[],
    sdkKey: string
  ): Promise<void> {
    await this.store.removeTargetAppsFromSDKKey(targetApps, sdkKey);
    await this.cache.clearSpecs(sdkKey);
    await this.cache.clearGlobalSDKKeys();
  }

  public async clearTargetAppsFromSDKKey(sdkKey: string): Promise<void> {
    await this.store.clearTargetAppsFromSDKKey(sdkKey);
    await this.cache.clearSpecs(sdkKey);
    await this.cache.clearGlobalSDKKeys();
  }

  public async getTargetAppNames(
    sdkKey?: string
  ): Promise<TargetAppNames | null> {
    if (sdkKey) {
      return this.store.getTargetAppsFromSDKKey(sdkKey);
    }
    return this.store.getTargetAppNames();
  }

  /* SDK Keys */

  public async registerSDKKey(sdkKey: string): Promise<void> {
    const sdkKeys = await this.store.addSDKKey(sdkKey);
    await this.cache.cacheSDKKeys(sdkKeys);
    await this.cache.clearGlobalSDKKeys();
  }

  public async deactivateSDKKey(sdkKey: string): Promise<void> {
    const sdkKeys = await this.store.removeSDKKey(sdkKey);
    await this.cache.cacheSDKKeys(sdkKeys);
    await this.cache.clearSpecs(sdkKey);
    await this.cache.clearGlobalSDKKeys();
  }

  public async getRegisteredSDKKeys(): Promise<Set<string>> {
    const cachedKeys = await this.cache.getSDKKeys();
    if (cachedKeys) {
      return cachedKeys;
    }

    const sdkKeys = await this.store.getRegisteredSDKKeys();
    await this.cache.cacheSDKKeys(sdkKeys);
    return sdkKeys;
  }

  public async getGlobalSDKKeys(): Promise<Set<string>> {
    const cached = await this.cache.getGlobalSDKKeys();
    if (cached) {
      return cached;
    }
    const sdkKeys = await this.getRegisteredSDKKeys();
    const globalKeys = new Set(
      await genFilter(Array.from(sdkKeys), async (sdkKey) => {
        const targetApps = await this.store.getTargetAppsFromSDKKey(sdkKey);
        return targetApps == null;
      })
    );
    await this.cache.cacheGlobalSDKKeys(globalKeys);
    return globalKeys;
  }
}
