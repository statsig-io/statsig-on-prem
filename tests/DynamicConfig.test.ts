import StatsigStorageExample from "../examples/StatsigStorageExample";
import StatsigOnPrem from "../src/StatsigOnPrem";
import StatsigSDK from "statsig-node";
import StatsigInstanceUtils from "statsig-node/dist/StatsigInstanceUtils";
import Evaluator from "statsig-node/dist/Evaluator";
import ConfigRuleBuilder from "../src/utils/ConfigRuleBuilder";
import { ConfigConditionType } from "../src/types/conditions/ConfigCondition";
import { APIOperatorType } from "../src/types/ConfigSpecs";
import { IDUtils } from "../src/utils/IDUtils";

describe("Dynamic Config", () => {
  const storage = new StatsigStorageExample();
  const statsig = new StatsigOnPrem(storage);
  const sdkKey = "secret-key";

  beforeEach(async () => {
    await statsig.initialize();
    await statsig.registerSDKKey(sdkKey);
  });

  afterEach(async () => {
    storage.clearAll();
    await statsig.clearCache();
    StatsigSDK.shutdown();
    jest.clearAllMocks();
  });

  it("Create Dynmaic Config", async () => {
    await statsig.createConfig("test-config", {
      defaultValue: { a_bool: true },
      enabled: true,
    });
    const configSpecs = await statsig.getConfigSpecs(sdkKey);
    await StatsigSDK.initialize("secret-key", {
      bootstrapValues: JSON.stringify(configSpecs),
    });
    const config = StatsigSDK.getConfigSync({ userID: "123" }, "test-config");
    expect(config.get("a_bool", false)).toEqual(true);
    StatsigSDK.shutdown();
  });

  it("Update Dynmaic Config", async () => {
    await statsig.createConfig("test-config", {
      defaultValue: { a_bool: true },
      enabled: true,
    });
    await statsig.updateConfig("test-config", {
      defaultValue: { a_bool: false },
    });
    const configSpecs = await statsig.getConfigSpecs(sdkKey);
    await StatsigSDK.initialize("secret-key", {
      bootstrapValues: JSON.stringify(configSpecs),
    });
    const config = StatsigSDK.getConfigSync({ userID: "123" }, "test-config");
    expect(config.get("a_bool", false)).toEqual(false);
  });

  it("Delete Dynmaic Config", async () => {
    await statsig.createConfig("test-config", {
      defaultValue: { a_bool: true },
      enabled: true,
    });
    await statsig.deleteConfig("test-config");
    const configSpecs = await statsig.getConfigSpecs(sdkKey);
    await StatsigSDK.initialize("secret-key", {
      bootstrapValues: JSON.stringify(configSpecs),
    });
    const StatsigSDKEvaluator = (StatsigInstanceUtils.getInstance() as any)
      ._evaluator as Evaluator;
    const evaluation = StatsigSDKEvaluator.getConfig(
      { userID: "123" },
      "test-config"
    );
    expect(evaluation.json_value).toEqual({});
    expect(evaluation.evaluation_details?.reason).toEqual("Unrecognized");
  });

  it("Target apps", async () => {
    await statsig.createConfig("test-config", {
      defaultValue: { a_bool: true },
      enabled: true,
    });
    let config = await statsig.getConfig("test-config");
    expect(config?.targetApps).toEqual(new Set([]));

    await statsig.addTargetAppsToConfig("test-config", ["target-app-1"]);
    config = await statsig.getConfig("test-config");
    expect(config?.targetApps).toEqual(new Set(["target-app-1"]));

    await statsig.removeTargetAppsFromConfig("test-config", ["target-app-1"]);
    config = await statsig.getConfig("test-config");
    expect(config?.targetApps).toEqual(new Set());
  });

  it("Update Dynmaic Config Rules", async () => {
    jest.spyOn(IDUtils, "generateNewSalt").mockReturnValue("mocked-salt");
    await statsig.createConfig("test-config", {
      defaultValue: { a_string: "default" },
      enabled: true,
    });
    const patchRules = (builder: ConfigRuleBuilder) => {
      builder.addRule({
        name: "Statsig Employee",
        conditions: [
          {
            type: ConfigConditionType.EMAIL,
            input: {
              operator: APIOperatorType.STR_CONTAINS_ANY,
              targetValue: ["@statsig.com"],
            },
          },
        ],
        value: { a_string: "statsig" },
        passPercentage: 100,
        idType: "userID",
      });
      const firstRule = builder.addRule({
        name: "Specific Users",
        conditions: [
          {
            type: ConfigConditionType.USER_ID,
            input: {
              operator: APIOperatorType.ANY,
              targetValue: ["123", "456"],
            },
          },
        ],
        value: { a_string: "user" },
        passPercentage: 100,
        idType: "userID",
      });
      builder.moveRuleUp(firstRule);
      return builder;
    };
    await statsig.updateConfig("test-config", { patchRules });
    const configSpecs = await statsig.getConfigSpecs(sdkKey);
    expect(configSpecs).toMatchSnapshot(
      { time: expect.any(Number) },
      "config-rules"
    );
    await StatsigSDK.initialize("secret-key", {
      bootstrapValues: JSON.stringify(configSpecs),
    });
    const config = StatsigSDK.getConfigSync(
      { userID: "123", email: "kenny@statsig.com" },
      "test-config"
    );
    expect(config.get("a_string", "default")).toEqual("user");
  });
});
