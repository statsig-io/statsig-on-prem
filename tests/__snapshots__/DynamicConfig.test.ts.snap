// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynamic Config Update Dynmaic Config Rules: config-rules 1`] = `
{
  "dynamic_configs": [
    {
      "defaultValue": {
        "a_string": "default",
      },
      "enabled": true,
      "entity": "dynamic_config",
      "idType": "userID",
      "isActive": undefined,
      "isDeviceBased": false,
      "name": "test-config",
      "rules": [
        {
          "conditions": [
            {
              "additionalValues": null,
              "field": "userID",
              "idType": "userID",
              "operator": "any",
              "targetValue": [
                "123",
                "456",
              ],
              "type": "user_field",
            },
          ],
          "id": "3272161690",
          "idType": "userID",
          "name": "Specific Users",
          "passPercentage": 100,
          "returnValue": {
            "a_string": "user",
          },
          "salt": "mocked-salt",
        },
        {
          "conditions": [
            {
              "additionalValues": null,
              "field": "email",
              "idType": "userID",
              "operator": "str_contains_any",
              "targetValue": [
                "@statsig.com",
              ],
              "type": "user_field",
            },
          ],
          "id": "634500529",
          "idType": "userID",
          "name": "Statsig Employee",
          "passPercentage": 100,
          "returnValue": {
            "a_string": "statsig",
          },
          "salt": "mocked-salt",
        },
      ],
      "salt": "mocked-salt",
      "type": "dynamic_config",
    },
  ],
  "feature_gates": [],
  "has_updates": true,
  "hashed_sdk_keys_to_entities": {},
  "layer_configs": [],
  "layers": {},
  "time": Any<Number>,
}
`;
